consultation_analyser/consultations/models.py

```diff
@@ -40,301 +41,237 @@ class Meta:
+class SlugFromTextModel(models.Model):
+    text = models.CharField(max_length=256)
+    slug = models.SlugField(null=False, editable=False, max_length=256)
 
-    class Meta(UUIDPrimaryKeyModel.Meta, TimeStampedModel.Meta):
-        constraints = [
-            models.UniqueConstraint(fields=["slug"], name="unique_consultation_slug"),
-        ]
+    def save(self, *args, **kwargs):
+        # Generate a slug from the text - ensure unique by adding timestamp if needed.
+        # Don't allow empty slug.
+        ModelClass = self.__class__
+        cropped_length = 220
+        cropped_text = self.text[:cropped_length]
+        generated_slug = slugify(cropped_text)
+        if self.pk:
+            slug_exists = (
+                ModelClass.objects.filter(slug=generated_slug).exclude(pk=self.pk).exists()
+            )
+        else:
+            slug_exists = ModelClass.objects.filter(slug=generated_slug).exists()
+        timestamp = datetime.datetime.now().strftime("%Y%m%d-%H%M%S%f")
+        if not generated_slug:
+            generated_slug = timestamp
+        elif slug_exists:
+            generated_slug = f"{generated_slug}-{timestamp}"
+        self.slug = generated_slug
+        return super().save(*args, **kwargs)
 
-    def delete(self, *args, **kwargs):
-        """
-        Delete Related theme objects that can become orphans
-        because deletes will not cascade from Answer because
-        the Theme in question could still be associated with
-        another Answer
-        """
-        Theme.objects.filter(answer__question__section__consultation=self).delete()
+    class Meta:
+        abstract = True
 
-        super().delete(*args, **kwargs)
 
-    @property
-    def latest_processing_run(self):
-        processing_runs = ProcessingRun.objects.filter(consultation=self).order_by("created_at")
-        latest = processing_runs.last() if processing_runs else None
-        return latest
+class Consultation(UUIDPrimaryKeyModel, TimeStampedModel, SlugFromTextModel):
+    users = models.ManyToManyField(User)
 
-    def get_processing_run(self, processing_run_slug=None):
-        # If None, get latest run if exists
-        if processing_run_slug:
-            processing_run = ProcessingRun.objects.get(consultation=self, slug=processing_run_slug)
-        else:
-            processing_run = self.latest_processing_run
-        return processing_run
+    class Meta(UUIDPrimaryKeyModel.Meta, TimeStampedModel.Meta, SlugFromTextModel.Meta):
+        constraints = [
+            models.UniqueConstraint(fields=["slug"], name="unique_consult_slug"),
+        ]
 
 
-class Section(UUIDPrimaryKeyModel, TimeStampedModel):
+class Question(UUIDPrimaryKeyModel, TimeStampedModel, SlugFromTextModel):
+    text = models.TextField()
     consultation = models.ForeignKey(Consultation, on_delete=models.CASCADE)
-    name = models.TextField()
-    slug = models.CharField(null=False, max_length=256)
+    order = models.IntegerField(null=True)
```

Comment:
Couple of questions: should these be unique for a given consultation? And do we want them to be nullable?

Reply:
followup question, do we want to overwrite `ordering` so that it's by order rather than `created_at`?

Reply:
Actually yes, they should be unique and non-nullable. Now I think about it more, why would you want your questions to be in a random order?

Didn't think of that - overwriting ordering sounds good.

Reply:
This field is now called `number` (raised by Duncan elsewhere), and we now have constraints so they are unique up to consultation. 

And I've rewritten the ordering to order by number.

Reply:
https://github.com/i-dot-ai/consult/pull/405/commits/517f14391a6c9008a953ce9152e36ae3e20adb92 - change `order` to `number`, make unique and non-nullable, order on this field (for `Question` and `QuestionPart`).

Reply:
Great! Thanks for putting all the related changes in one commit (much easier to follow).